discovery_engine/test/assets/data_provider_test.dart

```diff
@@ -0,0 +1,128 @@
+
+import 'utils/local_asset_server.dart' show LocalAssetServer;
+import 'utils/mock_http_asset_fetcher.dart' show HttpAssetFetcherWithCounter;
+import 'utils/mock_manifest_reader.dart'
+    show MockManifestReader, goodJson, wrongChecksumJson;
+
+void main() {
+  group('DataProvider', () {
+    group('getSetupData', () {
+      final outputPath = '${Directory.current.path}/test/assets/utils/output';
+      final baseAssetPath = '$outputPath/assets';
+      final vocabPath = '$baseAssetPath/smbert_v0000/vocab.txt';
+      final modelPath = '$baseAssetPath/smbert_v0000/smbert.onnx';
+      final assetFetcher = HttpAssetFetcherWithCounter('http://localhost:8080');
+      final manifestReader = MockManifestReader(goodJson);
+
+      late LocalAssetServer server;
+
+      setUpAll(() async {
+        server = await LocalAssetServer.start();
+      });
+
+      tearDown(() {
+        assetFetcher.resetCount();
+        Directory(outputPath).deleteSync(recursive: true);
+      });
+
+      tearDownAll(() {
+        server.close();
+      });
+
+      test(
+          'when provided with proper json manifest it can download assets '
+          'and asset fragments, and save them to a specified output path',
+          () async {
+        final dataProvider =
+            createDataProvider(assetFetcher, manifestReader, outputPath);
+
+        final setupData =
+            (await dataProvider.getSetupData()) as NativeSetupData;
+
+        expect(setupData.smbertVocab, equals(vocabPath));
+        expect(File(vocabPath).existsSync(), isTrue);
+        expect(setupData.smbertModel, equals(modelPath));
+        expect(File(modelPath).existsSync(), isTrue);
+      });
+
+      test(
+          'when the assets were already downloaded and the checksums are matching '
+          'it will serve those assets instead of fetching them again',
+          () async {
+        final dataProvider =
+            createDataProvider(assetFetcher, manifestReader, outputPath);
+        await _prepareOutputFiles(assetFetcher, manifestReader, baseAssetPath);
+
+        await dataProvider.getSetupData();
+
+        expect(assetFetcher.callCount, equals(0));
+      });
+
+      test(
+          'when the assets were already downloaded but the checksums '
+          'are NOT matching, it will fetch new files from the server',
+          () async {
+        final manifestReader = MockManifestReader(wrongChecksumJson);
+        final dataProvider =
+            createDataProvider(assetFetcher, manifestReader, outputPath);
+        await _prepareOutputFiles(assetFetcher, manifestReader, baseAssetPath);
+
+        await dataProvider.getSetupData();
+
+        expect(assetFetcher.callCount, equals(4));
+      });
+
+      test(
+          'when server responds with "503 - Service Unavailable" status '
+          'the fetcher is able to retry the request', () async {
+        await server.close();
+        server = await LocalAssetServer.start(retryCount: 1);
```

Comment:
I'm not sure if we can rely on execution order of async tests, or on them not interleaving.

If not this would affect other tests.

Reply:
I've changed the implementation, so it doesn't require restarting the local server, but anyways it shouldn't matter too much, as the test are executed in order specified in the file (as long as we don't [shuffle them](https://pub.dev/packages/test#shuffling-tests)), and for async tests the test runner won't consider a tast finished until the returned Future completes.